File: client/src/main/java/screen/AbstractPreferencesPanel.java
package screen;

import achievement.Reward;
import screen.preference.PreferencesDialog;
import util.Registry;

import javax.swing.*;
import java.awt.event.ActionListener;

public abstract class AbstractPreferencesPanel extends JPanel
											   implements ActionListener
{
	protected PreferencesDialog parent = null;
	
	public abstract void initVariables();
	public abstract boolean valid();
	public abstract void savePreferences();
	
	protected void toggleLockedComponent(AbstractButton c, Reward requiredReward)
	{
		boolean unlocked = requiredReward.isUnlocked();
		String text = unlocked? c.getText() : "Locked";
		String toolTipText = unlocked? c.getToolTipText() : "Unlock at " + requiredReward.getThreshold() + " achievements";
		
		c.setText(text);
		c.setToolTipText(toolTipText);
		c.setEnabled(unlocked);
	}
	
	public void setParent(PreferencesDialog parent)
	{
		this.parent = parent;
	}
}

File: client/src/main/kotlin/screen/preference/AbstractPreferencesPanel.kt
package screen.preference

import achievement.Reward
import java.awt.event.ActionListener
import javax.swing.AbstractButton
import javax.swing.JPanel

abstract class AbstractPreferencesPanel(protected val parentDialog: PreferencesDialog) :
    JPanel(), ActionListener {
    abstract fun initVariables()

    abstract fun valid(): Boolean

    abstract fun savePreferences()

    protected fun toggleLockedComponent(c: AbstractButton, requiredReward: Reward) {
        val unlocked = requiredReward.isUnlocked()
        val text = if (unlocked) c.text else "Locked"
        val toolTipText =
            if (unlocked) c.toolTipText
            else "Unlock at " + requiredReward.threshold + " achievements"

        c.setText(text)
        c.setToolTipText(toolTipText)
        c.setEnabled(unlocked)
    }
}
